src/wallet/rpcwallet.cpp

```diff
@@ -2567,6 +2652,401 @@ UniValue fundrawtransaction(const UniValue& params, bool fHelp)
+    CTransaction txBTC;
+    try {
+        ssTx >> txBTC;
+    }
+    catch (...) {
+        throw JSONRPCError(RPC_TYPE_ERROR, "The included bitcoinTx is malformed. Are you sure that is the whole string?");
+    }
+
+    std::vector<unsigned char> txOutProofData = ParseHex(params[2].get_str());
+    CDataStream ssTxOutProof(txOutProofData, SER_NETWORK, PROTOCOL_VERSION | SERIALIZE_BITCOIN_BLOCK_OR_TX);
+    CMerkleBlock merkleBlock;
+    try {
+        ssTxOutProof >> merkleBlock;
+    }
+    catch (...) {
+        throw JSONRPCError(RPC_TYPE_ERROR, "The included txoutproof is malformed. Are you sure that is the whole string?");
+    }
+    if (!ssTxOutProof.empty() || !CheckBitcoinProof(merkleBlock.header))
+        throw JSONRPCError(RPC_INVALID_PARAMETER, "Invalid tx out proof");
+
+    vector<uint256> txHashes;
+    vector<unsigned int> txIndices;
+    if (merkleBlock.txn.ExtractMatches(txHashes, txIndices) != merkleBlock.header.hashMerkleRoot)
+        throw JSONRPCError(RPC_INVALID_PARAMETER, "Invalid tx out proof");
+
+    if (txHashes.size() != 1 || txHashes[0] != txBTC.GetHash())
+        throw JSONRPCError(RPC_INVALID_PARAMETER, "The txoutproof must contain bitcoinTx and only bitcoinTx");
+
+    //Call contracthashtool
+    unsigned char nonce[16];
+    memset(nonce, 0, sizeof(nonce));
+    unsigned char fullcontract[40];
+    CScript mainchain_script = GetScriptForDestination(calculate_contract(Params().GetConsensus().fedpegScript, sidechainAddress, &nonce[0], fullcontract));
+
+    unsigned int nOut = 0;
+    for (; nOut < txBTC.vout.size(); nOut++)
+        if (txBTC.vout[nOut].scriptPubKey == mainchain_script)
+            break;
+    if (nOut == txBTC.vout.size())
+        throw JSONRPCError(RPC_INVALID_PARAMETER, "Failed to find output in bitcoinTx to the mainchain_address from getpeginaddress");
+    CAmount value = txBTC.vout[nOut].nValue.GetAmount();
+
+    uint256 genesisBlockHash = Params().ParentGenesisBlockHash();
+
+    //Output we're re-locking
+    CScript relock_spk;
+    relock_spk << std::vector<unsigned char>(genesisBlockHash.begin(), genesisBlockHash.end());
+    relock_spk << OP_WITHDRAWPROOFVERIFY;
+
+    if (value > MAX_MONEY / 200)
+        throw JSONRPCError(RPC_VERIFY_REJECTED, "IsStandard rules prevent pegging-in > 0.105 million BTC reliably at a time - please work with your functionary to mine a large lock-merge transaction first");
+
+    //Pad the locked outputs by the IsStandard lock dust value
+    CTxOut dummyTxOut(0, relock_spk);
+    CAmount lockDust(dummyTxOut.GetDustThreshold(withdrawLockTxFee));
+
+    LOCK(cs_main);
+
+    std::vector<std::pair<COutPoint, CAmount> > lockedUTXO;
+    if (!GetLockedOutputs(genesisBlockHash, value+lockDust, lockedUTXO))
+        throw JSONRPCError(RPC_INVALID_PARAMETER, "Failed to find inputs with sufficient value - are you sure bitcoinTx is valid?");
+
+    while (lockedUTXO.size() != 1) {
```

Comment:
`GetLockedOutputs` will return true IFF enough unspent unlocked funds were found. This should pretty much never return false(?) as it will spend in-mempool locked utxos as well, as a last resort. If it returns true that guarantees a non-0 amount of utxos.

>Are we creating a chain of transactions that consolidates locked UTXOs to one output on the sidechain?
Indeed. There are two types of OP_WPV spends(simplification):
1) Claims, which consume a single locked utxo and output a normal utxo
2) reductions, which take 2 utxos and combine them into 1

So if the value returned is spread among 5 utxos it will make a chain of 4 reductions, then a pegin transaction of the correct amount.